File: comparison/gram-winoground.R
library(tidyverse)
library(here)
library(glue)
library(latex2exp)
library(reticulate)
source("comparison/stats-helper.R")

np <- import("numpy")
oc_dir <- here("evals/gram-winoground/openclip/")
# oc_files <- list.files(oc_dir)

## make human data
get_human_data_wg <- function(manifest_file = "assets/gram-winoground/manifest.csv",
                              data_file = "evals/gram-winoground/human.jsonl") {
  included_trials <- read_csv(manifest_file) |> 
    mutate(trial = as.numeric(str_extract(image1, "(?<=ex_)[0-9]+(?=_img)")) + 1) |> 
    pull(trial)
  human_data <- read_lines(data_file) |> 
    str_remove_all('\\{\\"label\\"\\: \\"') |> 
    str_replace_all('\\", \\"score\\"\\: ', ',') |> 
    str_remove_all('\\}') |> 
    I() |> 
    read_csv(col_names = c("label", "score")) |> 
    separate_wider_delim(label, delim = "_", names = c("trial", "text", "image")) |> 
    mutate(trial = as.numeric(trial) + 1,
           image = (str_sub(image, 2) |> as.numeric()) + 1,
           text = (str_sub(text, 2) |> as.numeric()) + 1,
           pair = glue("image{image}text{text}")) |> 
    select(-image, -text) |> 
    filter(trial %in% included_trials) |> 
    mutate(trial = sapply(trial, \(t) which(included_trials == t)))
}

human_data_wg <- get_human_data_wg()

## comparison fxn
compare_wg <- function(model_data, human_data) {
  human_data_by_text <- human_data |> 
    separate_wider_regex(cols = pair,
                         patterns = c(image = "(?:image[12])", 
                                      text = "(?:text[12])")) |> 
    pivot_wider(names_from = image,
                values_from = score) |> 
    mutate(rowsum = rowSums(across(starts_with("image"))),
           across(starts_with("image"), \(x) x / rowsum),
           trial = glue("{trial}_{text}")) |> 
    select(-rowsum)
  
  model_data_by_text <- model_data |> 
    pivot_longer(cols = starts_with("image"),
                 names_to = c("image", "text"),
                 names_pattern = "(image[12])(text[12])",
                 values_to = "score") |> 
    pivot_wider(names_from = image,
                values_from = score) |> 
    mutate(trial = glue("{trial}_{text}"))
  
  opt_kl = get_opt_kl(human_data_by_text, model_data_by_text)
  tibble(kl = opt_kl$objective,
         beta = opt_kl$solution,
         iters = opt_kl$iterations)
}

## comparisons for openclip
openclip_div_wg <- lapply(oc_files, \(ocf) {
  epoch <- str_match(ocf, "[0-9]+")[1] |> as.numeric()
  res <- np$load(here(oc_dir, ocf)) |> 
    as_tibble() |> 
    `colnames<-`(value = c("image1text1", "image2text1", "image1text2", "image2text2")) |> 
    mutate(trial = seq_along(image1text1))
  kls <- compare_wg(res, human_data_wg) |> 
    mutate(epoch = epoch)
}) |> bind_rows()

wg_oc <- ggplot(openclip_div_wg, aes(x = log(epoch), y = kl)) +
  geom_point() +
  geom_smooth(col = my_palette[5]) +
  scale_colour_continuous() +
  # theme_classic() +
  labs(x = "log(Epoch)",
       y = TeX("Model–human dissimilarity ($D^*_{KL}$)"))

ggsave("comparison/gram-wg-oc.png", 
       wg_oc, 
       width = 6.2, height = 4.2, units = "in")

# mod_wg <- lm(cor ~ log(epoch), data = openclip_cors) |> 
#   tidy()

## comparisons for other models
wg_files <- c(list.files("evals/gram-winoground", pattern = "*.npy"), "openclip/openclip_epoch_256.npy")

other_res_wg <- lapply(wg_files, \(wgf) {
  res <- np$load(here("evals/gram-winoground", wgf)) |> 
    as_tibble()
  res <- res |> 
    `colnames<-`(value = c("image1text1", "image2text1", "image1text2", "image2text2")) |> 
    mutate(trial = seq_along(image1text1))
  acc <- res |> 
    mutate(correct = ((image1text1 > image1text2) + (image2text2 > image2text1))/2) |> 
    summarise(accuracy = mean(correct)) |> 
    pull(accuracy)
  kls <- compare_wg(res, human_data_wg) |> 
    mutate(model = wgf |> str_remove_all("wg_") |> str_remove_all(".npy"),
           accuracy = acc)
}) |> bind_rows() |> 
  mutate(model = str_replace_all(model, model_rename))

wg_all <- ggplot(other_res_wg, 
                 aes(x = accuracy, y = kl, shape = model)) + 
  geom_point() + 
  scale_shape_manual(values = c(16, 1, 17, 15, 18, 0, 2, 3)) +
  # theme_classic() +
  labs(x = "Accuracy",
       y = TeX("Model–human dissimilarity ($D^*_{KL}$)"),
       shape = "Model")

ggsave("comparison/gram-wg-all.png", 
       wg_all, 
       width = 6.2, height = 4.2, units = "in")

wg_feats <- other_res_wg |> 
  left_join(model_feats, by = join_by(model == Model)) |> 
  mutate(n_params = str_replace_all(`# params`, size_fix) |> as.numeric(),
         n_images = str_replace_all(`# images`, size_fix) |> as.numeric()) |> 
  summarise(accuracy = cor(kl, accuracy),
            size = cor(kl, log(n_params)),
            training = cor(kl, log(n_images)))

## item-level comparisons
item_res_wg <- lapply(wg_files, \(wgf) {
  res <- np$load(here("evals/gram-winoground", wgf)) |> 
    as_tibble()
  res <- res |> 
    `colnames<-`(value = c("image1text1", "image2text1", "image1text2", "image2text2")) |> 
    mutate(trial = seq_along(image1text1)) |> 
    pivot_longer(cols = starts_with("image"),
                 names_to = c("image", "text"),
                 names_pattern = "(image[12])(text[12])",
                 values_to = "score") |> 
    pivot_wider(names_from = image,
                values_from = score) |> 
    mutate(trial = glue("{trial}_{text}"))
  model_name = wgf |> str_remove_all("wg_") |> str_remove_all(".npy") |> 
    str_replace_all(model_rename)
  beta = other_res_wg |> filter(model == model_name) |> pull(beta)
  human_data_nested <- human_data_wg |> 
    separate_wider_regex(cols = pair,
                         patterns = c(image = "(?:image[12])", 
                                      text = "(?:text[12])")) |> 
    pivot_wider(names_from = image,
                values_from = score) |> 
    mutate(rowsum = rowSums(across(starts_with("image"))),
           across(starts_with("image"), \(x) x / rowsum),
           trial = glue("{trial}_{text}")) |> 
    select(-rowsum) |> 
    select(trial, starts_with("image"))
  kl <- get_mean_kl_img(human_data_nested, softmax_images(res, beta), return_distribs = TRUE) |> 
    # select(-distribs) |> 
    mutate(kl_z = scale(kl)[,1]) |> 
    mutate(model = model_name)
}) |> bind_rows()

item_dif_wg <- item_res_wg |> 
  separate_wider_delim(trial, delim = "_", names = c("trial", "text")) |> 
  group_by(trial) |> 
  summarise(kl_z = mean(kl_z))
  

File: comparison_py/gram-winoground.py
import os
import re
import io
from glob import glob
import numpy as np
import pandas as pd
from stats_helper import *

WG_DIR = "evals/gram-winoground/"

def get_human_data_wg(manifest_file="assets/gram-winoground/manifest.csv",
                      data_file="evals/gram-winoground/human.jsonl"):
    manifest_df = pd.read_csv(manifest_file)
    
    manifest_df['trial'] = manifest_df['image1'].str.extract(r"ex_([0-9]+)_img").astype(int) + 1
    included_trials = manifest_df['trial'].tolist()
    
    with open(data_file, 'r') as f:
        lines = f.readlines()
    processed_lines = [re.sub(r'[{}"]', '', line).replace('label: ', '').replace('score: ', '').strip() for line in lines]
    human_data = pd.read_csv(io.StringIO('\n'.join(processed_lines)), names=["label", "score"])
    
    human_data[['trial', 'text', 'image']] = human_data['label'].str.split('_', expand=True)
    human_data['trial'] = human_data['trial'].astype(int) + 1
    human_data['image'] = human_data['image'].str[1:].astype(int) + 1
    human_data['text'] = human_data['text'].str[1:].astype(int) + 1
    human_data['pair'] = human_data.apply(lambda row: f"image{row['image']}text{row['text']}", axis=1)
    human_data = human_data[human_data['trial'].isin(included_trials)]
    human_data['trial'] = human_data['trial'].apply(lambda t: included_trials.index(t) + 1)
    
    return human_data[['trial', 'pair', 'score']]

human_data_wg = get_human_data_wg()

def compare_wg(model_data, human_data):
    # process human data
    human_data_by_text = human_data.copy()
    human_data_by_text[['image', 'text']] = human_data_by_text['pair'].str.extract(r'(image[12])(text[12])')
    human_data_by_text = human_data_by_text.pivot(index=['trial', 'text'], columns='image', values='score').reset_index()
    human_data_by_text['rowsum'] = human_data_by_text.filter(like='image').sum(axis=1)
    for col in human_data_by_text.filter(like='image').columns:
        human_data_by_text[col] = human_data_by_text[col] / human_data_by_text['rowsum']
    human_data_by_text['trial'] = human_data_by_text.apply(lambda row: f"{row['trial']}_{row['text']}", axis=1)
    human_data_by_text = human_data_by_text.drop(columns=['rowsum', 'text'])
    
    # process model_data
    model_data_by_text = model_data.melt(id_vars=['trial'], value_vars=model_data.filter(like='image').columns,
                                         var_name='image_text', value_name='score')
    model_data_by_text[['image', 'text']] = model_data_by_text['image_text'].str.extract(r'(image[12])(text[12])')
    model_data_by_text = model_data_by_text.pivot(index=['trial', 'text'], columns='image', values='score').reset_index()
    model_data_by_text['trial'] = model_data_by_text.apply(lambda row: f"{row['trial']}_{row['text']}", axis=1)
    model_data_by_text = model_data_by_text.drop(columns=['text'])

    opt_kl = get_opt_kl(human_data_by_text, model_data_by_text)    
    result_df = pd.DataFrame({
        'kl': opt_kl['objective'],
        'beta': opt_kl['solution'],
        'iterations': opt_kl['iterations']
    }, index=[0])
    return result_df

# comparisons for openclip
oc_files = glob(WG_DIR + "openclip/*.npy")
oc_kls = []
for ocf in oc_files:
    if not os.path.exists(ocf):
        continue
    res = np.load(ocf)
    res = pd.DataFrame(res.reshape((-1, 4)),
                       columns = ["image1text1", "image1text2", "image2text1", "image2text2"]) # order is different than R
    res['trial'] = np.arange(res.shape[0])+1
    res['correct'] = ((res['image1text1'] > res['image1text2']).astype(int) + \
                      (res['image2text2'] > res['image2text1']).astype(int))/2
    acc = res['correct'].mean()
    kls = compare_wg(res, human_data_wg)
    kls['epoch'] = int(os.path.splitext(os.path.basename(ocf))[0].replace("openclip_epoch_", ""))
    kls['accuracy'] = acc
    oc_kls.append(kls)
if len(oc_kls) > 0:
    openclip_div_wg = pd.concat(oc_kls).sort_values(["epoch"]).reset_index(drop=True)
    openclip_div_wg.to_csv("comparison/gram-wg_openclip.csv")

# comparisons for other models
wg_files = glob(WG_DIR + "*.npy") + [WG_DIR + "openclip/openclip_epoch_256.npy"]
all_kls = []
for wgf in wg_files:
    if not os.path.exists(wgf):
        continue
    res = np.load(wgf)
    res = pd.DataFrame(res.reshape((-1, 4)),
                       columns = ["image1text1", "image1text2", "image2text1", "image2text2"]) # order is different than R
    res['trial'] = np.arange(res.shape[0])+1
    res['correct'] = ((res['image1text1'] > res['image1text2']).astype(int) + \
                      (res['image2text2'] > res['image2text1']).astype(int))/2
    acc = res['correct'].mean()
    kls = compare_wg(res, human_data_wg)
    kls['model'] = os.path.splitext(os.path.basename(wgf))[0].replace("wg_", "").replace("_epoch_256", "")
    kls['accuracy'] = acc
    print(kls)
    all_kls.append(kls)
other_res_wg = pd.concat(all_kls).sort_values(["model"]).reset_index(drop=True)
other_res_wg.to_csv("comparison/gram-wg_models.csv")
